inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp

```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+    float bbox2Size = bboxSizes[idx2];
+
+    return intersectSize / (bbox1Size + bbox2Size - intersectSize);
+}
+
+void MKLDNNDetectionOutputNode::NMSCF(int* indicesIn,
+                                        int& detections,
+                                        int* indicesOut,
+                                        const float* bboxes,
+                                        const float* boxSizes) {
+    // nms for this class
+    int countIn = detections;
+    detections = 0;
+    for (int i = 0; i < countIn; ++i) {
+        const int prior = indicesIn[i];
 
         bool keep = true;
         for (int k = 0; k < detections; ++k) {
-            const int kept_idx = indices[k];
-            float overlap = JaccardOverlap(bboxes, sizes, idx, kept_idx);
-            if (overlap > _nms_threshold) {
+            const int keptPrior = indicesOut[k];
+            float overlap = JaccardOverlap(bboxes, boxSizes, prior, keptPrior);
+            if (overlap > NMSThreshold) {
                 keep = false;
                 break;
             }
         }
         if (keep) {
-            indices[detections] = idx;
+            indicesOut[detections] = prior;
             detections++;
         }
     }
 }
 
-void MKLDNNDetectionOutputNode::nms_mx(const float* conf_data,
-                                 const float* bboxes,
-                                 const float* sizes,
-                                 int* buffer,
-                                 int* indices,
-                                 int* detections,
-                                 int num_priors_actual) {
-    int count = 0;
-    for (int i = 0; i < num_priors_actual; ++i) {
-        float conf = -1;
-        int id = 0;
-        for (int c = 1; c < _num_classes; ++c) {
-            float temp = conf_data[c*_num_priors + i];
-            if (temp > conf) {
-                conf = temp;
-                id = c;
-            }
-        }
-
-        if (id > 0 && conf >= _confidence_threshold) {
-            indices[count++] = id*_num_priors + i;
-        }
-    }
-
-    int num_output_scores = (_top_k == -1 ? count : (std::min)(_top_k, count));
-
-    std::partial_sort_copy(indices, indices + count,
-                           buffer, buffer + num_output_scores,
-                           ConfidenceComparator(conf_data));
-
-    for (int i = 0; i < num_output_scores; ++i) {
-        const int idx = buffer[i];
-        const int cls = idx/_num_priors;
-        const int prior = idx%_num_priors;
-
+void MKLDNNDetectionOutputNode::NMSMX(int* indicesIn,
+                                    int* detections,
+                                    int* indicesOut,
+                                    const float* bboxes,
+                                    const float* sizes) {
+    // Input is candidate for image, output is candidate for each class within image
+    int countIn = detections[0];
+    detections[0] = 0;
+
+    for (int i = 0; i < countIn; ++i) {
```

Comment:
Could we use parallelism?

Reply:
NMS routine can not paralleled because the next iteration depends on previous iteration which may change the gradually generated output. The next iteration need all current outputs to suppress. Thank you.

Reply:
Yes, you're right. Thank you for the explanation!